devito/function.py

```diff
@@ -259,43 +275,137 @@ def _indices(cls, **kwargs):
             dimensions = grid.dimensions
         return dimensions
 
+    def _allocate_memory(func):
+        """Allocate memory as a :class:`Data`."""
+        def wrapper(self):
```

Comment:
Interesting approach. I would have gone for making `self._data` a `cached_property`, but I guess this works too.

Reply:
Oh yes, that works too, but then you cannot initialise ``self._data`` to ``None`` in the constructor (right?) 

I agree a ``cached_property`` would work, but perhaps this makes it more explicit the fact that we are allocating on demand

Reply:
Sure, sounds plausible.